### PR TITLE
GenericAuthControl - authenticate without need to create custom Control

### PR DESCRIPTION
**Pull request by MartinMystikJonas**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "nette/http": "^3.0.5"
   },
   "require-dev": {
+    "nette/application": "^v3.0.0",
```
Comment (f3l1x) [inline composer.json:24]:
```suggestion
    "nette/application": "^3.0.0",
```

**Comment by MartinMystikJonas** [comment]
Dependencies fixed, rebased on master and fixed BC with older Nette packages used with --prefer-lowest (Arrays::invoke is not available till nette/utils 3.2.1)

Now it should be ready to merge.

File: tests/fixtures/Presenter/TestPresenter.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+use Nette\Http\Response;
+use Nette\Http\UrlScript;
+
+final class TestPresenter extends Presenter
+{
+
+	/** @var Component */
+	private $component;
+
+	/** @var Request */
```
Comment (f3l1x) [inline tests/fixtures/Presenter/TestPresenter.php:17]: HttpRequest

File: tests/fixtures/Presenter/TestPresenter.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	public $httpRequest;
+
+	/** @var Response */
```
Comment (f3l1x) [inline tests/fixtures/Presenter/TestPresenter.php:20]: HttpRequest alias

File: tests/fixtures/Presenter/TestPresenter.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	public $httpResponse;
+
+	/**	@var \Nette\Application\Response */
```
Comment (f3l1x) [inline tests/fixtures/Presenter/TestPresenter.php:23]: FQN

File: tests/fixtures/Presenter/TestPresenter.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		$this->component = $component;
+		$this->getComponent('subject');
+	}
+
+	protected function createComponentSubject(): Component
+	{
+		return $this->component;
+	}
+
+	public function sendResponse(\Nette\Application\Response $response): void
```
Comment (f3l1x) [inline tests/fixtures/Presenter/TestPresenter.php:45]: FQN

File: src/UI/Components/GenericAuthControl.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+use Nette\Bridges\ApplicationLatte\Template;
+use UnexpectedValueException;
+
+class GenericAuthControl extends Control
+{
+
+  /** @var AuthCodeFlow */
+	private $authCodeFlow;
+
+  /** @var string|null */
```
Comment (f3l1x) [inline src/UI/Components/GenericAuthControl.php:19]: Formatting

**Pull request merged**